jina/drivers/multimodal.py

```diff
@@ -0,0 +1,114 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from collections import defaultdict
+from typing import Iterable, Tuple, Dict
+
+import numpy as np
+
+from .encode import BaseEncodeDriver
+from .helper import pb2array, array2pb
+from ..proto import jina_pb2
+
+
+def _extract_doc_content(doc: 'jina_pb2.Document'):
+    """Returns the content of the document with the following priority:
+    If the document has an embedding, return it, otherwise return its content.
+    """
+    if doc.embedding.buffer:
+        return pb2array(doc.embedding)
+    else:
+        return doc.text or doc.buffer or (doc.blob and pb2array(doc.blob))
+
+
+def _extract_modalities_from_document(doc: 'jina_pb2.Document'):
+    """Returns a dictionary of document content (embedding, text, blob or buffer) with `modality` as its key
+    """
+    doc_content = {}
+    for chunk in doc.chunks:
+        modality = chunk.modality
+        if modality in doc_content:
+            return None
```

Comment:
i dont get this early return, what's the purpose here @bwanglzu @JoanFM

Reply:
This None will make it an invalid document similar to what happens in the EncodeDriver. It works under the assumption that every chunk should have different modality.

Reply:
>  every chunk should have different modality

can you elaborate more on this? Say we have a PDF document, then the first granularity is
```
Doc
 |
 -- Chunk1 (image)
 |
 -- Chunk2 (text)
```
This assumption holds true only at this granularity, if we continue, this is not true

```
Doc
 |
 -- Chunk1 (image)
          |
          |-- ChunkChunk1 (image)
          |-- ChunkChunk2 (image)
 |
 -- Chunk2 (text)
```

Reply:
it is true, but the user can have a previous multimodal driver to merge the `ChunkChunk1` and `ChunkChunk2` into `Chunk1`.